### PR TITLE
AO3-6467 Track deletion of previous fnok on update

### PR DESCRIPTION
**Pull request by ceithir**
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6467?focusedCommentId=360436

## Purpose

Add the missing log when a fnok is replaced (akin to one deletion and one addition) 

## Testing Instructions

1. Add AccountB as an FNOK for AccountA. Check it shows in both the logs of AccountA and AccountB
2. Update the FNOK for AccountA to be AccountC. Check the deletion of previous fnok shows in the logs of AccountA and AccountB, and the addition of the new fnok shows in the log of AccountA and AccountC.

## Credit

Ceithir (he/him)

### CODE REVIEW

File: app/controllers/admin/admin_users_controller.rb [hunk 1]
```diff
@@ -87,9 +87,18 @@ def update_next_of_kin
       return
     end
 
+    previous_fnok_user_id = fnok&.kin&.id
     fnok = @user.build_fannish_next_of_kin if fnok.blank?
     fnok.assign_attributes(kin: kin, kin_email: kin_email)
     if fnok.save
+      if previous_fnok_user_id
+        @user.create_log_item({
```
Comment (weeklies) [inline app/controllers/admin/admin_users_controller.rb:95]: This logic is very similar to the case at line 76, I am wondering if we can simplify the method by refactoring.

**Comment by ceithir** [comment]
Note: I'll be about fully AFK for one week starting tomorrow. So, depending on how urgent this piece of code is (I'm not fully clear on the release schedule), someone else might need to pick up the torch of this PR.

**Comment by weeklies** [comment]
I'm away from my computer, so this isn't tested at all but how about: 

(I also added `if fnok.changed?` as I notice a change was being logged when an admin presses update on a user with an FNOK without changing anything.)

```suggestion
  def update_next_of_kin
    @user = authorize User.find_by!(login: params[:user_login])
    fnok = @user.fannish_next_of_kin
    kin = User.find_by(login: params[:next_of_kin_name])
    kin_email = params[:next_of_kin_email]
return if kin == fnok.kin && kin_email == fnok.kin_email

    if fnok.present? 
      fnok.destroy 
      @user.create_log_item({
                              action: ArchiveConfig.ACTION_ADD_FNOK,
                              fnok_user_id: fnok.kin.id,
                              admin_id: current_admin.id,
                              note: "Change made by #{current_admin.login}"
                            })

      if kin.blank? && kin_email.blank?
        flash[:notice] = ts("Fannish next of kin was removed.") 
        redirect_to admin_user_path(@user)
        return 
      end
    end 

    fnok = @user.build_fannish_next_of_kin if fnok.blank?
    fnok.assign_attributes(kin: kin, kin_email: kin_email)
    if fnok.changed? && fnok.save
      flash[:notice] = ts("Fannish next of kin was updated.")
      redirect_to admin_user_path(@user)
    else
      @hide_dashboard = true
      log_items
      render :show
    end
  end
```

**Comment by ceithir** [comment]
Tackling the issue of updating the kin to itself rather than dancing around it like I've been doing is indeed likely good.

On the other hand, not a super fan of deleting the previous kin before validating the form data, which only happens in the `fnok.save` below unless I've missed something.

**Comment by weeklies** [comment]
Good point, I can take a second crack on the weekend if this isn't merged by then.

File: app/controllers/admin/admin_users_controller.rb [hunk 2]
```diff
@@ -69,27 +69,33 @@ def update
...
     @user = authorize User.find_by!(login: params[:user_login])
     fnok = @user.fannish_next_of_kin
+    previous_fnok_user_id = fnok&.kin&.id
     kin = User.find_by(login: params[:next_of_kin_name])
     kin_email = params[:next_of_kin_email]
 
     if kin.blank? && kin_email.blank?
-      if fnok.present?
+      if fnok.blank?
+        flash[:notice] = ts("No change to Fannish next of kin.")
```
Comment (sarken) [inline app/controllers/admin/admin_users_controller.rb:78]: Just noting "Fannish" should be lowercase here and on line 90. 

Taking another shot at this this weekend is fine!

File: spec/controllers/admin/admin_users_controller_spec.rb [hunk 3]
```diff
@@ -259,6 +259,53 @@
       expect(log_item.admin_id).to eq(admin.id)
       expect(log_item.note).to eq("Change made by #{admin.login}")
     end
+
+    it "logs updating a fannish next of kin" do
+      admin = create(:support_admin)
```
Comment (weeklies) [inline spec/controllers/admin/admin_users_controller_spec.rb:264]: I notice there are repeated lines of 
```
      admin = create(:support_admin)
      fake_login_admin(admin)
```

We could surround the FNOK tests with a shared context to be more efficient.

```
    context "when admin has support role" do
      let(:admin) { create(:support_admin) }
```

File: app/controllers/admin/admin_users_controller.rb [hunk 4]
```diff
@@ -69,27 +69,33 @@ def update
...
+    end
+
+    if fnok&.kin == kin && fnok&.kin_email == kin_email
+      flash[:notice] = ts("No change to Fannish next of kin.")
       redirect_to admin_user_path(@user)
       return
     end
 
     fnok = @user.build_fannish_next_of_kin if fnok.blank?
     fnok.assign_attributes(kin: kin, kin_email: kin_email)
```
Comment (weeklies) [inline app/controllers/admin/admin_users_controller.rb:96]: Second shot at improving the update logic.

```suggestion
  def update_next_of_kin
    @user = authorize User.find_by!(login: params[:user_login])
    kin = User.find_by(login: params[:next_of_kin_name])
    kin_email = params[:next_of_kin_email]

    fnok = @user.fannish_next_of_kin
    previous_fnok_user_id = fnok&.kin&.id
    fnok ||= @user.build_fannish_next_of_kin
    fnok.assign_attributes(kin: kin, kin_email: kin_email)

    unless fnok.changed?
      flash[:notice] = ts("No change to fannish next of kin.")
      redirect_to admin_user_path(@user)
      return
    end

    # Remove FNOK that already exists.
    if fnok.persisted? && kin.blank? && kin_email.blank?
      fnok.destroy
      log_next_of_kin_removed(previous_fnok_user_id)
      flash[:notice] = ts("Fannish next of kin was removed.")
      redirect_to admin_user_path(@user)
      return
    end
```

This code would require a test for invalid updates to be changed. As the notice "No change to fannish next of kin." will not be shown, but rather the error message from the validation, "Kin email can't be blank"

```
      it "does nothing if fnok is kept undefined" do
        kin = build(:fannish_next_of_kin, user: user, kin_email: "")

        expect(kin).not_to be_valid
        expect(kin.errors[:kin_email]).to include("can't be blank")
        expect(user.reload.log_items).to be_empty
      end
```

File: app/controllers/admin/admin_users_controller.rb [hunk 5]
```diff
@@ -184,4 +190,15 @@ def activate
   def log_items
     @log_items ||= (@user.log_items + LogItem.where(fnok_user_id: @user.id)).sort_by(&:created_at).reverse
   end
+
+  def log_next_of_kin_removed(user_id)
```
Comment (weeklies) [inline app/controllers/admin/admin_users_controller.rb:194]: Would marking this as `private` be helpful?

**Review by sarken — APPROVED** [review]

**Pull request merged**